core/domain/email_services.py

```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+            'Malformed sender email address: %s' % sender_email)
+
+    bcc = feconf.ADMIN_EMAIL_ADDRESS if (bcc_admin) else ''
+    reply_to = (
+        get_incoming_email_address(reply_to_id)
+        if reply_to_id else '')
+
+    if not constants.DEV_MODE:
+        platform_email_services.send_email_to_recipients(
+            sender_email, [recipient_email], subject.encode(encoding='utf-8'),
+            plaintext_body.encode(encoding='utf-8'),
+            html_body.encode(encoding='utf-8'), bcc=[bcc], reply_to=reply_to)
+    else:
+        msg_title = 'MailgunService.SendMail'
+        # pylint: disable=division-operator-used
+        msg_body = (
+            """
+            From: %s
+            To: %s
+            Subject: %s
+            Body:
+                Content-type: text/plain
+                Data length: %d
+            Body
+                Content-type: text/html
+                Data length: %d
+            """ % (
+                sender_email, recipient_email, subject, len(plaintext_body),
+                len(html_body)))
+        # pylint: enable=division-operator-used
+        msg = msg_title + dedent(msg_body)
+        logging.info(msg)
+        logging.info(
+            'You are not currently sending out real email since this is a' +
+            ' dev environment. Emails are sent out in the production' +
+            ' environment.')
+
+
+def send_bulk_mail(
+        sender_email, recipient_emails, subject, plaintext_body, html_body):
+    """Sends an email using mailgun api.
+
+    In general this function should only be called from
+    email_manager._send_email().
+
+    Args:
+        sender_email: str. the email address of the sender. This should be in
+            the form 'SENDER_NAME <SENDER_EMAIL_ADDRESS>'.
+        recipient_emails: list(str). list of the email addresses of recipients.
+        subject: str. The subject line of the email.
+        plaintext_body: str. The plaintext body of the email.
+        html_body: str. The HTML body of the email. Must fit in a datastore
+            entity.
+
+    Raises:
+        Exception: if the configuration in feconf.py forbids emails from being
+            sent.
+        Exception: if mailgun api key is not stored in feconf.MAILGUN_API_KEY.
+        Exception: if mailgun domain name is not stored in
+            feconf.MAILGUN_DOMAIN_NAME.
+            (and possibly other exceptions, due to mail.send_mail() failures)
+    """
+    if not feconf.CAN_SEND_EMAILS:
+        raise Exception('This app cannot send emails to users.')
+
+    for recipient_email in recipient_emails:
+        if not is_email_valid(recipient_email):
+            raise ValueError(
+                'Malformed recipient email address: %s' % recipient_email)
+
+    if not is_sender_email_valid(sender_email):
+        raise ValueError(
+            'Malformed sender email address: %s' % sender_email)
+
+    # To send bulk emails we pass list of recipients in 'to' paarameter of
```

Comment:
This stuff, which is platform-dependent and is likely to vary between different platforms, should be handled at the platform layer.

Reply:
Done Migrated over